share_app/src/main/java/org/odk/share/tasks/DownloadJob.java

```diff
@@ -178,42 +192,77 @@ private void readForm() {
-            receiveFile(FORM_PATH);
+            String formName = receiveFile(FORMS_PATH);
             int numOfRes = dis.readInt();
+            String formMediaPath = FORMS_PATH + "/" + displayName + "-media";
             while (numOfRes-- > 0) {
-                receiveFile(FORM_PATH + displayName + "-media");
+                receiveFile(formMediaPath);
             }
+
+            // Add row in forms db
+            ContentValues values = new ContentValues();
+            values.put(FormsProviderAPI.FormsColumns.FORM_FILE_PATH, FORMS_PATH + "/" + formName);
+            values.put(FormsProviderAPI.FormsColumns.DISPLAY_NAME, displayName);
+            values.put(FormsProviderAPI.FormsColumns.JR_FORM_ID, formId);
+            values.put(FormsProviderAPI.FormsColumns.JR_VERSION, formVersion);
+            values.put(FormsProviderAPI.FormsColumns.SUBMISSION_URI, submissionUri);
+            values.put(FormsProviderAPI.FormsColumns.FORM_MEDIA_PATH, formMediaPath);
+            new FormsDao().saveForm(values);
         } catch (IOException e) {
             Timber.e(e);
         }
     }
 
-    private void readInstances(String formId) {
+    private void readInstances(String formId, String formVersion) {
         try {
             int numInstances = dis.readInt();
             while (numInstances-- > 0) {
                 // publish current progress
                 rxEventBus.post(new DownloadEvent(DownloadEvent.Status.DOWNLOADING, ++progress, total));
+                String displayName = dis.readUTF();
+                String submissionUri = dis.readUTF();
+
+                if (submissionUri.equals("-1")) {
+                    submissionUri = null;
+                }
 
                 int numRes = dis.readInt();
                 String time = new SimpleDateFormat("yyyy-MM-dd_HH-mm-ss-SSS",
                         Locale.ENGLISH).format(Calendar.getInstance().getTime());
-                String path = INSTANCE_PATH + formId + "_" + time;
+                String path = INSTANCES_PATH + "/" + formId + "_" + time;
                 while (numRes-- > 0) {
                     receiveFile(path);
                 }
+
+                // Add row in instances table
+                ContentValues values = new ContentValues();
+                values.put(DISPLAY_NAME, displayName);
+                values.put(STATUS, InstanceProviderAPI.STATUS_COMPLETE);
+                values.put(CAN_EDIT_WHEN_COMPLETE, "true");
+                values.put(SUBMISSION_URI, submissionUri);
+                values.put(INSTANCE_FILE_PATH, path);
+                values.put(JR_FORM_ID, formId);
+                values.put(JR_VERSION, formVersion);
+                Uri uri = new InstancesDao().saveInstance(values);
+
+                // Add row in share table
+                ContentValues shareValues = new ContentValues();
+                shareValues.put(INSTANCE_ID, Long.parseLong(uri.getLastPathSegment()));
```

Comment:
I'm storing info about the instances, and for forms transfer, I'm not keeping any track of `forms`, if the forms are received then I'm just updating the `forms` table. Is it required to track the forms also?

Reply:
> Why are you calling getLastPathSegment()

I'm just storing the instanceID, which was added to a `instances` table.
`getLastPathSegment()` extracts the id of the newly added row. I'm not storing the `uri`.

Reply:
Got it, that makes sense. Since instances are specific records it's important to know who has had access to them when. A form with a particular form_id can be duplicated a bunch of times and it doesn't matter so much where they come from. So that works.

What about the name `INSTANCE_ID`?

Reply:
> getLastPathSegment() extracts the id of the newly added row

Great, that makes sense. I haven't had a chance to try this myself because I don't have two devices at the moment. @shobhitagarwal1612's screenshot confused me because all the instanceID values are showing up as single digits.